Adds `IO.getModifiedTimeOrZero(file)` and `IO.setModifiedTimeOrFalse(file)`

cunei opened the pull request:
Ref https://github.com/sbt/io/pull/92

There are just too many instances in which sbt's code relies on
the `lastModified`/`setLastModified` semantics, so instead of moving
to `get`/`setModifiedTime`, we use new IO calls that offer the new
timestamp precision, but retain the old semantics.

Commit:
Add IO.lastModified and IO.setLastModified, improve scaladoc

Commit:
Revert *ModifiedTime() calls to *lastModified*() calls

There are just too many instances on which sbt's code relies on
the lastModified/setLastModified semantics, so instead of moving
to get/setModifiedTime, we use new IO calls that offer the new
timestamp precision, but retain the old semantics.

Commit:
Minor comment change

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?io=pull/106/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

Commit:
Small optimization for MacMilli

The attr descriptor can be reused, so there is no need to
reallocate and reinitialize it every time.

eed3si9n commented on io/src/main/scala/sbt/io/IO.scala:
```diff
@@ -1197,4 +1172,67 @@ object IO {
+   *
+   * @see getModifiedTime
+   */
+  def lastModified(file: File): Long =
```
I think it's too subtle/confusing to have different variety of `getModifiedTime` that differs in similar sounding name.

My suggestion here would be to have two versions:
- `getModifiedTimeTry` that returns `scala.util.Try`.
- `getModifiedTime` that returns `Long` and uses `0L` when the file doesn't exist.

cunei replied:
`IO.lastModified` and `IO.setLastModified` have the exact same semantics as `f.lastModified` and `f.setLastModified`, which is why retaining the same names is important.
The other methods are just ancillary, and after these patches they will be mostly unused within sbt.
Actually, we have spent a great deal of time already on what should have been a minor matter, so I'd avoid further tweaks. At this point I would strongly recommend that we just merge these pull requests, and move forward with other topics.

eed3si9n replied:
Any public method on `IO` becomes semi-permanent API of sbt, so if we can avoid creating confusing duplicates that will save our headaches in the future. Since io 1.1.2 is already out with `IO.getModifiedTime(..)` we are committed with the name. For the same reason we are going through this exercise, anyone who tries to replace the usage of `File#lastModified` should use the return-`0L` version.

Given all other modules are already calling `IO.getModifiedTime(..)`, changing here also saves changing the code downstream.

cunei replied:
Since `IO.getModifiedTime` is already included in `IO`, we have to keep respecting the specification described in the Scaladoc. It is already described as a method that throws exception, so we cannot change it again now.

If you are concerned about too many time-related methods being exposed in `IO`, it is perfectly ok to only expose `xxxModifiedTime`, keeping the `xxxlastModified` calls internal. I can just place them in `internal.io.Milli`. However, we cannot change `IO.getModifiedTime` again, now.

Consequently, the only possible course of action is transitioning the downstream code to (differently named) calls that go back to the `0L` semantics. That is what these pull requests do, upon @dwijnand's suggestion.

It is convenient to name them in the same way as the `java.util.File` ones, for clarity. It is true however that they do not need to be exposed in `IO` at all. (Incidentally, you are the one who asked for `getModifiedTime` to be exposed in `IO`)

So, as a middle ground solution, I propose that I remove `lastModified` and `setLastModified` from `IO` and move them to `internal.io.Milli`, changing downstream accordingly.

eed3si9n replied:
Given that sbt 1.1.0 hasn't reached final, I am personally ok with `IO.getModifiedTime` changing the semantics from the previous description in Scaladoc.

I am fine with either:
- Change to return `IO.getModifiedTime` `0L`, or
- Make `xxxLastModified` internal.

@dwijnand What do you think?

dwijnand replied:
Given this is sbt/io and not sbt I'm of the opinion that we can't change semantics between stable releases (aka from 1.1.2 to 1.1.3).

What's a good alternative name to `getModifiedTime` to put on `IO` that means "last modified time or 0`. `getLastModifiedOrZero`? `getLastModifiedSafe`?

Personally I'm fine with `lastModified`, but I'd be equally happy if a better name is found that's distinct enough from `getModifiedTime` to avoid confusion.

eed3si9n replied:
How about we deprecate `IO.getModifiedTime` and add:

1. `getLastModifiedTry` that uses `Try`
2. `getLastModifiedOrZero` that returns `0L`

?

dwijnand replied:
Fine by me.

cunei replied:
Then you end up with *three* variants. Confusion is increased even more!... :smile:
We need the zero-returning variant, and we need to keep the exception-throwing variant since it's in the API. There is really no need to add the "Try" one as well...

eed3si9n replied:
If it's deprecated then it's effectively out of the API.
`Try` should be the replacement for someone who wants throwy behavior.

dwijnand replied:
Maybe instead of a `getLastModifiedTry` it should be `getLastModifiedUnsafe` which throws exceptions.

cunei commented:
We had a long discussion concerning these pull requests, and developed a compromise that, although not particularly pretty in the interim, will eventually get us to a single set of calls with the `Try` semantics. In short:
- The zero-returning variants are problematic, and we will move away from them eventually, but not now
- The exception-throwing variants are better, but having the same prototype as the ones above may be confusing
- The `Try` version is preferable, but we cannot add it right away, otherwise we end up with too many calls in `IO` (get/set/copy, in three variants, would mean having nine different calls in `IO`)

So:

- We are going to stick with the zero-returning semantics for now (equivalent to `java.io.File`'s `lastModified`/`setLastModified`)
- In order to make it clear that the methods may deal with zero timestamps, however, in `IO` we will use for them the names `getModifiedTimeOrZero()` and `setModifiedTimeOrFalse()` (the existing `copyLastModified()`, already in the API, will be maintained)
- The just-added `get`/`set`/`copyModifiedTime` will be deprecated, so that they may be eventually removed.
- In a future version of sbt/io, once we remove the `modifiedTime` calls, we will add instead the versions that return `Try` (or `Either`), so that the different types can guide us into a migration. The zero-returning variants will be deprecated.

In conclusion, eventually we will have only three get/set/copy calls, based on `Try`.

Commit:
Merge branch '1.1.x' into wip-milliBis

Commit:
Change modifiedTime definitions

Commit:
Enable modified files test

Since https://github.com/sbt/io/issues/82 is now closed,
enable the previously pending test.

cunei commented:
@dwijnand, since #82 is closed I re-enabled the pending modified files detection test; please review.

Commit:
Remove warnings

eed3si9n reviewed (dismissed):
LGTM. Thanks for your patience on this Toni!

cunei commented:
@eed3si9n I added an additional warning removal commit and your approval went away; please approve again. Thanks!

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/337/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/3822/head-pull/470/head-pull/106/head-pull/192/head-pull/136//head-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/337/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/3822">pull/3822/head      </A> | sbt/sbt@94e36a14c80027a4e316505574d4c49fdcb14948
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/470">pull/470/head       </A> | sbt/zinc@67d387856b3a30c757af306e2672d98f8630459b
  io                | <A HREF="https://github.com/sbt/io/pull/106">pull/106/head       </A> | sbt/io@77d3ba2fd47aa24be7cff3666223253024dea7ff
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/pull/192">pull/192/head       </A> | sbt/librarymanagement@72dd0064c3791ad4a6a2ba1904d8cf12322aede3
  util              | <A HREF="https://github.com/sbt/util/pull/136/">pull/136//head      </A> | 
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/3822/head&zinc=pull/470/head&io=pull/106/head&librarymanagement=pull/192/head&util=pull/136//head&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/338/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/3822/head-pull/470/head-pull/106/head-pull/192/head-pull/136/head-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/338/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/3822">pull/3822/head      </A> | sbt/sbt@94e36a14c80027a4e316505574d4c49fdcb14948
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/470">pull/470/head       </A> | sbt/zinc@67d387856b3a30c757af306e2672d98f8630459b
  io                | <A HREF="https://github.com/sbt/io/pull/106">pull/106/head       </A> | sbt/io@77d3ba2fd47aa24be7cff3666223253024dea7ff
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/pull/192">pull/192/head       </A> | sbt/librarymanagement@72dd0064c3791ad4a6a2ba1904d8cf12322aede3
  util              | <A HREF="https://github.com/sbt/util/pull/136">pull/136/head       </A> | sbt/util@0a1bd5a3b25f95d25a5fef18844cb0acf61a425b
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/3822/head&zinc=pull/470/head&io=pull/106/head&librarymanagement=pull/192/head&util=pull/136/head&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

cunei commented:
@eed3si9n The five pull requests are ready for merging now.

dwijnand commented on io/src/test/scala/sbt/internal/io/SourceModificationWatchSpec.scala:
```diff
@@ -21,10 +21,9 @@ abstract class SourceModificationWatchSpec(
-    pending // until fixed https://github.com/sbt/io/issues/82
+    watchTest(parentDir) {
+      IO.write(file, "bar")
+    }
```
This shouldn't've been enabled like this because it fails on HFS+. See https://github.com/sbt/io/pull/102/files#diff-32eb449ec95851dd791288244207a8f8R32.
